core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -293,8 +293,15 @@ public void commit() {
             taskOps.commit(base, updated.withUUID());
           });
 
+    } catch (CommitFailedException commitFailedException) {
+      // We have an acknowledged failure from the Catalog. We are confident that the commit has not been applied
+      Exceptions.suppressAndThrow(commitFailedException, this::cleanAll);
     } catch (RuntimeException e) {
-      Exceptions.suppressAndThrow(e, this::cleanAll);
+      LOG.error("Cannot determine whether the commit was successful or not, the underlying data files may or " +
+              "may not be needed. Manual intervention via the Remove Orphan Files Action can remove these " +
+              "files when a connection to the Catalog can be re-established if the commit was actually unsuccessful. " +
```

Comment:
I think it will be helpful to advise the user to check whether their operation succeeded before retrying. If the commit was applied and the change was retried, this will result in duplicated data.

Reply:
Ideally, that should be part of the exception message.